File: bindings/java/src/test/java/ua/dstucrypto/dstucore/ThreadSafetyTest.java
package ua.dstucrypto.dstucore;

import org.junit.jupiter.api.Test;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.fail;

/**
 * T-219: concurrency contract for this binding's own wrapper types, decided and recorded rather
 * than assumed. Two shapes:
 * <ul>
 * <li><b>{@link Sign}'s static methods are safe to call concurrently on the same key bytes.</b>
 * Unlike .NET/C++, this binding's {@code Sign} holds no native handle at all - every key is a
 * plain Java {@code byte[]} copied into the native call each time (see {@code Sign.java}'s own
 * doc), so there is no shared mutable native state to race on in the first place. Verified below
 * by real concurrent JNI calls, not assumed from the API shape alone.</li>
 * <li><b>{@link SecretStreamPushState}/{@link SecretStreamPullState} are NOT safe to share across
 * threads</b> - each holds one native handle whose pointed-to state (nonce/counter) advances with
 * every {@code push}/{@code pull} call, with no lock anywhere in this wrapper. The supported
 * concurrency model is one stream per thread. Verified below: many threads, each driving its own
 * encrypt/decrypt pair concurrently, all round-trip correctly - deliberately not tested by racing
 * a single shared instance, since that would just induce undefined behavior on the native side
 * rather than test a contract.</li>
 * </ul>
 */
class ThreadSafetyTest {
    @Test
    void concurrentVerifyOnSharedKeyIsSafe() throws Exception {
        byte[] signingKey = Sign.keygen();
        byte[] verifyingKey = Sign.verifyingKey(signingKey);
        byte[] message = "shared-key concurrent verify".getBytes(StandardCharsets.US_ASCII);
        byte[] signature = Sign.sign(signingKey, message);

        int threadCount = 16;
        int perThread = 200;
        runConcurrently(threadCount, () -> {
            for (int i = 0; i < perThread; i++) {
                if (!Sign.verify(verifyingKey, message, signature)) {
                    fail("Sign.verify returned false on a valid signature");
                }
            }
            return null;
        });
    }

    @Test
    void concurrentSignOnSharedKeyIsSafe() throws Exception {
        byte[] signingKey = Sign.keygen();
        byte[] verifyingKey = Sign.verifyingKey(signingKey);
        byte[] message = "shared-key concurrent sign".getBytes(StandardCharsets.US_ASCII);

        int threadCount = 16;
        int perThread = 50;
        runConcurrently(threadCount, () -> {
            for (int i = 0; i < perThread; i++) {
                byte[] sig = Sign.sign(signingKey, message);
                if (!Sign.verify(verifyingKey, message, sig)) {
                    fail("a concurrently-produced signature failed to verify");
                }
            }
            return null;
        });
    }

    @Test
    void concurrentIndependentSecretstreamLoopsAreSafe() throws Exception {
        int threadCount = 8;
        int perThreadChunks = 20;

        runConcurrently(threadCount, threadIndex -> {
            byte[] key = SecretStream.keygen();
            byte[][] chunks = new byte[perThreadChunks][];
            for (int i = 0; i < perThreadChunks; i++) {
                chunks[i] = ("thread " + threadIndex + " chunk " + i).getBytes(StandardCharsets.US_ASCII);
            }

            ByteArrayOutputStream out = new ByteArrayOutputStream();
            try (SecretStreamEncryptor enc = new SecretStreamEncryptor(key, out)) {
                for (byte[] chunk : chunks) {
                    enc.write(chunk);
                }
                enc.complete();
            }

            ByteArrayOutputStream expected = new ByteArrayOutputStream();
            for (byte[] chunk : chunks) {
                expected.write(chunk);
            }

            try (SecretStreamDecryptor dec = new SecretStreamDecryptor(key, new ByteArrayInputStream(out.toByteArray()))) {
                assertArrayEquals(expected.toByteArray(), dec.readAll());
            }
            return null;
        });
    }

    /**
     * Runs {@code task} concurrently on {@code threadCount} real OS threads (a fixed thread pool,
     * not just async tasks that might serialize onto one carrier thread), each passed its own
     * index, and re-throws the first exception/assertion failure any of them hit.
     */
    private interface IndexedTask {
        Void call(int threadIndex) throws Exception;
    }

    private static void runConcurrently(int threadCount, Callable<Void> task) throws Exception {
        runConcurrently(threadCount, threadIndex -> task.call());
    }

    private static void runConcurrently(int threadCount, IndexedTask task) throws Exception {
        ExecutorService pool = Executors.newFixedThreadPool(threadCount);
        try {
            List<Future<Void>> futures = new ArrayList<>();
            for (int i = 0; i < threadCount; i++) {
                int threadIndex = i;
                futures.add(pool.submit(() -> task.call(threadIndex)));
            }
            for (Future<Void> future : futures) {
                try {
                    future.get(60, TimeUnit.SECONDS);
                } catch (ExecutionException e) {
                    if (e.getCause() instanceof Exception) {
                        throw (Exception) e.getCause();
                    }
                    throw e;
                }
            }
        } finally {
            pool.shutdown();
        }
    }
}

File: bindings/go/dstu/thread_safety_test.go
package dstu

import (
	"bytes"
	"fmt"
	"io"
	"sync"
	"testing"
)

// T-219: concurrency contract for this binding's own wrapper types, decided and recorded rather
// than assumed.
//
//   - Read-only key types are safe to share across goroutines. VerifyingKey/SigningKey wrap an
//     immutable native key behind a raw *C pointer; every operation on them (Verify, Sign) only
//     reads that key - no caller-visible mutable state exists to race on. Verified below by
//     calling the SAME *VerifyingKey/*SigningKey concurrently from many goroutines under
//     `go test -race`.
//   - Stateful streaming types are NOT safe to share across goroutines -
//     SecretStreamEncryptWriter/SecretStreamDecryptReader hold a native push/pull state that
//     advances (nonce/counter) with every call, with no lock anywhere in this wrapper. The
//     supported concurrency model is one stream per goroutine, each with its own instance.
//     Verified below: many goroutines, each driving its own encrypt/decrypt pair concurrently, all
//     round-trip correctly - deliberately not tested by racing a single shared instance, since that
//     would just induce undefined behavior on the native side rather than test a contract.

func TestConcurrentVerifyOnSharedKeyIsSafe(t *testing.T) {
	signingKey, err := GenerateSigningKey()
	if err != nil {
		t.Fatal(err)
	}
	defer signingKey.Close()
	verifyingKey := signingKey.VerifyingKey()
	defer verifyingKey.Close()

	message := []byte("shared-key concurrent verify")
	sig := signingKey.Sign(message)

	const goroutines = 16
	const perGoroutine = 200
	var wg sync.WaitGroup
	errs := make(chan error, goroutines)
	for i := 0; i < goroutines; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := 0; j < perGoroutine; j++ {
				ok, err := verifyingKey.Verify(message, sig)
				if err != nil {
					errs <- err
					return
				}
				if !ok {
					errs <- fmt.Errorf("Verify returned false on a valid signature")
					return
				}
			}
		}()
	}
	wg.Wait()
	close(errs)
	for err := range errs {
		t.Error(err)
	}
}

func TestConcurrentSignOnSharedKeyIsSafe(t *testing.T) {
	signingKey, err := GenerateSigningKey()
	if err != nil {
		t.Fatal(err)
	}
	defer signingKey.Close()
	verifyingKey := signingKey.VerifyingKey()
	defer verifyingKey.Close()

	message := []byte("shared-key concurrent sign")

	const goroutines = 16
	const perGoroutine = 50
	var wg sync.WaitGroup
	errs := make(chan error, goroutines)
	for i := 0; i < goroutines; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := 0; j < perGoroutine; j++ {
				sig := signingKey.Sign(message)
				ok, err := verifyingKey.Verify(message, sig)
				if err != nil {
					errs <- err
					return
				}
				if !ok {
					errs <- fmt.Errorf("a concurrently-produced signature failed to verify")
					return
				}
			}
		}()
	}
	wg.Wait()
	close(errs)
	for err := range errs {
		t.Error(err)
	}
}

func TestConcurrentIndependentSecretstreamLoopsAreSafe(t *testing.T) {
	const goroutines = 8
	const perGoroutineChunks = 20
	var wg sync.WaitGroup
	errs := make(chan error, goroutines)

	for g := 0; g < goroutines; g++ {
		wg.Add(1)
		go func(goroutineIndex int) {
			defer wg.Done()

			key, err := GenerateSecretstreamKey()
			if err != nil {
				errs <- err
				return
			}
			defer key.Close()

			chunks := make([][]byte, perGoroutineChunks)
			var expected bytes.Buffer
			for i := 0; i < perGoroutineChunks; i++ {
				chunks[i] = []byte(fmt.Sprintf("goroutine %d chunk %d", goroutineIndex, i))
				expected.Write(chunks[i])
			}

			var buf bytes.Buffer
			enc, err := NewSecretStreamEncryptWriter(&buf, key, true)
			if err != nil {
				errs <- err
				return
			}
			for _, chunk := range chunks {
				if _, err := enc.Write(chunk); err != nil {
					errs <- err
					return
				}
			}
			if err := enc.Complete(); err != nil {
				errs <- err
				return
			}
			if err := enc.Close(); err != nil {
				errs <- err
				return
			}

			dec, err := NewSecretStreamDecryptReader(&buf, key, true)
			if err != nil {
				errs <- err
				return
			}
			defer dec.Close()
			decrypted, err := io.ReadAll(dec)
			if err != nil {
				errs <- err
				return
			}
			if !bytes.Equal(expected.Bytes(), decrypted) {
				errs <- fmt.Errorf("goroutine %d: round trip mismatch", goroutineIndex)
			}
		}(g)
	}

	wg.Wait()
	close(errs)
	for err := range errs {
		t.Error(err)
	}
}
